src/FloatingMenuPlugin.ts

```diff
@@ -0,0 +1,570 @@
+            e.stopPropagation();
+
+            const pos = {
+              x: e ? e.clientX : 0,
+              y: e ? e.clientY : 0,
+            };
+
+            openFloatingMenu(plugin, view, undefined, undefined, pos);
+          }
+        });
+
+        // --- Close popup on outside click ---
+        const outsideClickHandler = (e: MouseEvent) => {
+          const el = e.target as HTMLElement;
+          if (
+            plugin._popUpHandle &&
+            !el.closest('.context-menu') &&
+            !el.closest('.float-icon')
+          ) {
+            plugin._popUpHandle.close(null);
+            plugin._popUpHandle = null;
+          }
+        };
+        document.addEventListener('click', outsideClickHandler);
+        return {};
+      },
+    });
+  }
+
+  getEffectiveSchema(schema: Schema): Schema {
+    return schema;
+  }
+}
+
+export function copySelectionRich(
+  view: EditorView,
+  plugin: FloatingMenuPlugin
+) {
+  const { state } = view;
+  if (state.selection.empty) return;
+
+  if (!view.hasFocus()) view.focus();
+
+  const slice = state.selection.content();
+
+  const sliceJSON = {
+    content: slice.content.toJSON(),
+    openStart: slice.openStart,
+    openEnd: slice.openEnd,
+    sliceModel: createSliceObject(view),
+  };
+
+  navigator.clipboard
+    .writeText(JSON.stringify(sliceJSON))
+    .then(() => { }) //console.log('Rich content copied')
+    .catch((err) => console.error('Clipboard write failed', err));
+  if (plugin._popUpHandle) {
+    plugin._popUpHandle.update({
+      ...plugin._popUpHandle['props'],
+      pasteAsReferenceEnabled: true,
+    });
+  }
+  if (plugin._popUpHandle?.close) {
+    plugin._popUpHandle.close(null);
+    plugin._popUpHandle = null;
+  }
+}
+
+export function createSliceObject(editorView: EditorView): SliceModel {
+  const instanceUrl = 'http://modusoperandi.com/editor/instance/';
+  const referenceUrl = 'http://modusoperandi.com/ont/document#Reference_nodes';
+
+  const sliceModel: SliceModel = {
+    name: '',
+    description: '',
+    id: '',
+    referenceType: '',
+    source: '',
+    from: '',
+    to: '',
+    ids: [],
+  };
+
+  const objectIds: string[] = [];
+  let firstParagraphText: string | null = null;
+
+  editorView.focus();
+
+  const $from = editorView.state.selection.$from;
+  const $to = editorView.state.selection.$to;
+
+  const from = $from.start($from.depth);
+  const to = $to.end($to.depth);
+
+  editorView.state.doc.nodesBetween(from, to, (node) => {
+    if (node.type.name === 'paragraph') {
+      if (!firstParagraphText && node.textContent?.trim()) {
+        firstParagraphText = node.textContent.trim();
+      }
+      if (node.attrs?.objectId) {
+        objectIds.push(node.attrs.objectId);
+      }
+    }
+  });
+
+  sliceModel.id = instanceUrl + uuidv4();
+  sliceModel.ids = objectIds;
+  sliceModel.from = objectIds.length > 0 ? objectIds[0] : '';
+  sliceModel.to = objectIds.length > 0 ? objectIds[objectIds.length - 1] : '';
```

Comment:
Sonar: Prefer `.at(…)` over `[….length - index]`.
```suggestion
  sliceModel.to = objectIds.length > 0 ? objectIds.at(-1) : '';
```